core/http/server-spring/src/test/java/org/eclipse/rdf4j/http/server/repository/transaction/TestActiveTransactionRegistry.java

```diff
@@ -0,0 +1,152 @@
+import org.eclipse.rdf4j.repository.RepositoryException;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TestActiveTransactionRegistry {
+
+	private static final Logger logger = LoggerFactory.getLogger(TestActiveTransactionRegistry.class);
+
+	private ActiveTransactionRegistry registry;
+
+	private RepositoryConnection conn;
+
+	private UUID txnId1;
+
+	private UUID txnId2;
+
+	/**
+	 * @throws java.lang.Exception
+	 */
+	@Before
+	public void setUp()
+		throws Exception
+	{
+		System.setProperty(ActiveTransactionRegistry.CACHE_TIMEOUT_PROPERTY, "1");
+		registry = ActiveTransactionRegistry.INSTANCE;
+		conn = Mockito.mock(RepositoryConnection.class);
+		txnId1 = UUID.randomUUID();
+		txnId2 = UUID.randomUUID();
+
+	}
+
+	@Test
+	public void testTimeoutRepeatedAccess()
+		throws Exception
+	{
+		registry.register(txnId2, conn);
+
+		int count = 0;
+		while (count++ < 2) {
+			logger.debug("pass {}", count);
+			registry.getTransactionConnection(txnId2);
+			Thread.sleep(1200);
+			registry.returnTransactionConnection(txnId2);
+		}
+
+		registry.deregister(txnId2);
+		try {
+			registry.getTransactionConnection(txnId2);
+			fail("should be deregistered");
+		}
+		catch (RepositoryException e) {
+			// fall through, expected
+		}
+	}
+
+	@Test
+	public void testMultithreadedAccess() {
+
+		CountDownLatch txn1registered = new CountDownLatch(1);
+
+		CountDownLatch done = new CountDownLatch(2);
+
+		Runnable r1 = new Runnable() {
+
+			@Override
+			public void run() {
+				registry.register(txnId1, conn);
+				txn1registered.countDown();
+
+				try {
+					registry.getTransactionConnection(txnId1);
+					Thread.sleep(700);
+					registry.returnTransactionConnection(txnId1);
+
+					done.countDown();
+				}
+				catch (RepositoryException | InterruptedException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				}
+			}
+		};
+
+		Runnable r2 = new Runnable() {
+
+			@Override
+			public void run() {
+				try {
+					txn1registered.await();
+				}
+				catch (InterruptedException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				}
+
+				try {
+					registry.getTransactionConnection(txnId1);
+					Thread.sleep(500);
+					registry.returnTransactionConnection(txnId1);
+
+					done.countDown();
+				}
+				catch (RepositoryException | InterruptedException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				}
+			}
+		};
+
+		Runnable r3 = new Runnable() {
+
+			@Override
+			public void run() {
+				try {
+					done.await();
+					registry.deregister(txnId1);
+				}
+				catch (InterruptedException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				}
+			}
+		};
+
+		Thread t1 = new Thread(r1, "r1");
+		Thread t2 = new Thread(r2, "r2");
+		Thread t3 = new Thread(r3, "r3");
+
+		t3.start();
+		t2.start();
+		t1.start();
```

Comment:
Done. See below: main thread now waits for thread 3 (which, in turn, waits for all other threads to complete).